common/src/main/java/org/apache/nemo/common/MetricUtils.java

```diff
@@ -0,0 +1,503 @@
+      "postgres", "fake_password")) {
+      try (final Statement statement = c.createStatement()) {
+        statement.setQueryTimeout(30);  // set timeout to 30 sec.
+
+        if (MUST_UPDATE_EP_KEY_METADATA.getCount() == 0) {
+          insertOrUpdateMetadata(c, "EP_KEY_METADATA", EP_KEY_METADATA);
+          LOG.info("EP Key Metadata saved to DB.");
+        }
+
+        if (MUST_UPDATE_EP_METADATA.getCount() == 0) {
+          insertOrUpdateMetadata(c, "EP_METADATA", EP_METADATA);
+          LOG.info("EP Metadata saved to DB.");
+        }
+      }
+    } catch (SQLException e) {
+      LOG.warn("Saving of Metadata to DB failed: ", e);
+    }
+  }
+
+  /**
+   * Utility method to save key, value to the metadata table.
+   * @param c the connection to the DB.
+   * @param key the key to write to the DB metadata table.
+   * @param value the value to write to the DB metadata table.
+   * @throws SQLException SQLException on the way.
+   */
+  public static void insertOrUpdateMetadata(final Connection c, final String key, final Serializable value)
+    throws SQLException {
+    try (final PreparedStatement pstmt = c.prepareStatement(
+      "INSERT INTO " + METADATA_TABLE_NAME + " (key, data) "
+        + "VALUES ('" + key + "', ?) ON CONFLICT (key) DO UPDATE SET data = excluded.data;")) {
+      pstmt.setBinaryStream(1,
+        new ByteArrayInputStream(SerializationUtils.serialize(value)));
+      pstmt.executeUpdate();
+    }
+  }
+
+  /**
+   * Stringify execution properties of an IR DAG.
+   * @param irdag IR DAG to observe.
+   * @return the pair of stringified execution properties. Left is for vertices, right is for edges.
+   */
+  public static Pair<String, String> stringifyIRDAGProperties(final IRDAG irdag) {
+    final StringBuilder vStringBuilder = new StringBuilder();
+    final StringBuilder eStringBuilder = new StringBuilder();
+
+    irdag.getVertices().forEach(v ->
+      v.getExecutionProperties().forEachProperties(ep ->
+        epFormatter(vStringBuilder, VERTEX, v.getNumericId(), ep)));
+
+    irdag.getVertices().forEach(v ->
+      irdag.getIncomingEdgesOf(v).forEach(e ->
+        e.getExecutionProperties().forEachProperties(ep ->
+          epFormatter(eStringBuilder, EDGE, e.getNumericId(), ep))));
+
+    // Update the metric metadata if new execution property key / values have been discovered and updates are required.
+    updateMetaData();
+    return Pair.of(vStringBuilder.toString().trim(), eStringBuilder.toString().trim());
+  }
+
+  /**
+   * Formatter for execution properties. It updates the metadata for the metrics if new EP key / values are discovered.
+   * @param builder string builder to append the metrics to.
+   * @param idx index specifying whether it's a vertex or an edge. This should be one digit.
+   * @param numericId numeric ID of the vertex or the edge.
+   * @param ep the execution property.
+   */
+  private static void epFormatter(final StringBuilder builder, final int idx,
```

Comment:
I saw the similar code before and now it's gone and comes with this new PR.
What happens??

Reply:
https://github.com/apache/incubator-nemo/pull/192
I've found the PR.

Reply:
The file has been moved and changed, it looks like git doesn't recognize them as small changes, and simply marks the file as a new file. 😢 I'm unsure how to deal with this without fixing git itself.